File: app/src/main/java/com/example/pc/basemvp/base/BaseFragment.java
package com.example.pc.basemvp.base;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

public abstract class BaseFragment extends Fragment {

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        return this.initRootView(inflater, container, savedInstanceState);
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        this.bundleData(savedInstanceState);
        this.loadData(savedInstanceState);
    }

    protected abstract View initRootView(@NonNull LayoutInflater inflater,
            @Nullable ViewGroup container, @Nullable Bundle savedInstanceState);

    protected abstract void bundleData(Bundle savedInstanceState);

    protected abstract void loadData(Bundle savedInstanceState);
}

File: app/src/main/java/com/example/flowmvvm/base/BaseFragment.kt
package com.example.flowmvvm.base

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleCoroutineScope
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.viewbinding.ViewBinding
import com.example.flowmvvm.utils.extension.createViewModelLazy
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import kotlin.reflect.KClass

abstract class BaseFragment<B : ViewBinding, VM : BaseViewModel> : Fragment() {

    protected abstract val viewModelClass: KClass<VM>
    val viewModel: VM by createViewModelLazy(viewModelClass)

    protected abstract val bindingInflater: (LayoutInflater, ViewGroup?, Boolean) -> B
    private var _binding: B? = null
    val binding get() = _binding!!

    protected val coroutineScope: LifecycleCoroutineScope
        get() = viewLifecycleOwner.lifecycleScope

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = bindingInflater.invoke(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        bindViewModel()
        setupView()
        bindView()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun bindViewModel() {
        viewModel.tag = viewModelClass.simpleName.toString()
        launchStart { viewModel.loadingEvent.collectLatest { setLoading(it) } }
        launchStart { viewModel.errorEvent.collectLatest { it?.let { onHandleError(it) } } }
    }

    fun setLoading(isShow: Boolean) {
        val parentActivity = (requireActivity() as BaseActivity<*, *>)
        parentActivity.setLoading(isShow)
    }

    fun onHandleError(throwable: Throwable) {
        val parentActivity = (requireActivity() as BaseActivity<*, *>)
        parentActivity.onHandleError(throwable)
    }

    fun launchStart(block: suspend () -> Unit) {
        with(viewLifecycleOwner) {
            lifecycleScope.launch {
                lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
                    block.invoke()
                }
            }
        }
    }

    protected abstract fun setupView()

    protected abstract fun bindView()
}
